From 564272e8649cc13d4e2e7b63839ac5de8cc5a3c4 Mon Sep 17 00:00:00 2001
From: Guo Mengqi <guomengqi3@huawei.com>
Date: Thu, 3 Nov 2022 06:41:57 +0000
Subject: [PATCH] mm/sharepool: check size=0 in mg_sp_make_share_k2u()
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 564272e8649cc13d4e2e7b63839ac5de8cc5a3c4
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5QQPG
CVE: NA

--------------------------------

Add a size-0-check in mg_sp_make_share_k2u() to avoid passing 0-size spa
to __insert_sp_area().

Signed-off-by: Guo Mengqi <guomengqi3@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 mm/share_pool.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index f1f29bc07361..09e0b247bddb 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -2870,6 +2870,11 @@ static int sp_k2u_prepare(unsigned long kva, unsigned long size,
 	unsigned int page_size = PAGE_SIZE;
 	unsigned long kva_aligned, size_aligned;
 
+	if (!size) {
+		pr_err_ratelimited("k2u input size is 0.\n");
+		return -EINVAL;
+	}
+
 	if (sp_flags & ~SP_FLAG_MASK) {
 		pr_err_ratelimited("k2u sp_flags %lx error\n", sp_flags);
 		return -EINVAL;
-- 
2.33.0

